knc now uses poll() instead of select()

gstrauss opened the pull request:
- poll with POLLRDHUP when POLLRDUP is supported
  - save POLLRDHUP if received, cease reading once socket buffers empty
  - poll without timeout (-1; INFTIM) when POLLRDHUP is supported
- check for and return if not waiting for any events (sockets closed)
- check that shut_nread_lwrite, shut_nwrite_lread run no more than once
- propagate effects of close() if shutdown() fails
- retry (later) on read() or write() if EINTR or EAGAIN

gstrauss committed:
knc now uses poll() instead of select()

- poll with POLLRDHUP when POLLRDUP is supported
  - save POLLRDHUP if received, cease reading once socket buffers empty
  - poll without timeout (-1; INFTIM) when POLLRDHUP is supported
- check for and return if not waiting for any events (sockets closed)
- propagate effects of close() if shutdown() fails

nicowilliams commented on bin/knc.c:
```diff
@@ -47,6 +55,10 @@
 #include <unistd.h>
 
+#ifndef POLLRDHUP
+#define POLLRDHUP 0
```
How about:

```
#ifndef POLRDHUP
#define POLLRDHUP POLLHUP
#endif
```

gstrauss replied:
POLLHUP should be ignored in pfd->events, but might theoretically result in EINVAL on some platforms.

gstrauss replied:
On Linux, man page suggests POLLHUP is ignored in events.  NetBSD man page documents that POLLHUP is checked even if not present in events, implying that it is accepted in events.  Ok.  I'll make that change, but also have to add a preprocessor check around setting the timeout for poll, since a timeout is suggested if POLLRDHUP is not supported.

gstrauss replied:
According to the spec, EWOULDBLOCK can be defined and different from EAGAIN, so I can not take that suggestion.  I'll add a check for EWOULDBLOCK for portability.

nicowilliams replied:
Interesting -- the Open Group (POSIX, SUS) manpage says nothing about that.  The Linux one says that `POLLHUP` is meaningless in `events`.  Fine.

gstrauss replied:
Old reference to differences between poll() implementations on different unix and unix-like platforms: https://www.greenend.org.uk/rjk/tech/poll.html

gstrauss replied:
Please note that the reason that the existing code makes assumptions that it is called to move data only if 'work' is in a state that is ready to accept additional data.  I preserved that behavior to minimize the scope of code changes.  Therefore, I only check for POLLIN, which should only be returned if it was requested in pfd->events.  That's why I am not checking for (POLLIN|POLLRDHUP|POLLHUP) before read() so that read() detects EOF when the socket buffers are empty  Intead, I store a flag that POLLRDHUP|POLLHUP was received.

nicowilliams commented on bin/knc.c:
```diff
@@ -47,6 +55,10 @@
+#ifndef POLLRDHUP
+#define POLLRDHUP 0
+#endif
+
```
Maybe:

```
#ifndef EWOULDBLOCK
#define EWOULDBLOCK EAGAIN
#endif
```

then check only for `EWOULDBLOCK` in the code.  That would maximize portability.

gstrauss replied:
According to the spec, EWOULDBLOCK can be defined and different from EAGAIN, so I can not take that suggestion. I'll add a check for EWOULDBLOCK for portability.

I pushed another commit adding checks for EWOULDBLOCK in addition to EAGAIN.

nicowilliams commented on bin/knc.c:
```diff
@@ -795,7 +794,10 @@ move_local_to_network_buffer(work_t *work)
 	} else if (work->network_buffer.in_len < 0) {
-		if (errno == ECONNRESET)
+		const int errnum = errno;
+		if (errnum == EINTR || errnum == EAGAIN)
```
And here check for `EWOULDBLOCK` per above comment.

gstrauss replied:
added check in subsequent commit

nicowilliams commented on bin/knc.c:
```diff
@@ -795,7 +794,10 @@ move_local_to_network_buffer(work_t *work)
 		return 0;
 	} else if (work->network_buffer.in_len < 0) {
-		if (errno == ECONNRESET)
+		const int errnum = errno;
```
We don't need this `errnum`.  We only need this when we'll call things like `close(2)` that could stomp on the current value of `errno`, and when we do my preference is to call that `save_errno`.

gstrauss replied:
As you said, it is not needed in these cases to preserve errno.  It is an idiom I have seen used and used myself for a long time to store errno in errnum for reuse, rather than repeating a function call to obtain the (often) thread-local errno value.

Would you prefer that I repeat the use of errno rather than storing in errnum?

gstrauss replied:
changed to repeat use of errno.

nicowilliams commented on bin/knc.c:
```diff
@@ -846,8 +848,11 @@ write_local_buffer(work_t *work)
-		if (errno == EPIPE) {
+	if (len < 0) {
+		const int errnum = errno;
+		if (errnum == EINTR || errnum == EAGAIN) {
```
Ditto.

nicowilliams commented on bin/knc.c:
```diff
@@ -955,8 +960,11 @@ write_network_buffer(work_t *work)
+	if (len < 0) {
+		const int errnum = errno;
+		if (errnum == EINTR || errnum == EAGAIN) {
+			return 1; /* retry later (do not return <= 0) */
```
Ditto.

nicowilliams commented on bin/knc.c:
```diff
@@ -999,40 +1007,41 @@ write_network_buffer(work_t *work)
+	struct pollfd * const pfd_net = pfds+0;
+	struct pollfd * const pfd_in  = pfds+1;
+	struct pollfd * const pfd_out = pfds+2;
+	struct pollfd * const pfd_err = pfds+3;
```
Why the `const`?

gstrauss replied:
It is an indication to the reader that those values are not going to change in the struct pollfd list pfds[] in how this routine uses that list.  Also, it is a (minor) compiler hint to the optimizer.

nicowilliams commented on bin/knc.c:
```diff
@@ -999,40 +1007,41 @@ write_network_buffer(work_t *work)
+	pfd_in->fd  = work->local_in;
+	pfd_out->fd = work->local_out;
+	pfd_err->fd = work->local_err;
+	nfds = (-1 == work->local_err) ? 3 : 4;
```
Clearly it's not Roland's style to put the literal on the left of the comparison operator.

Since modern C compilers warn about assignments in conditionals, putting the literal on the left is not that useful for avoiding accidental assignment typos.

gstrauss replied:
moved literal to right-hand-side in two places

nicowilliams commented on bin/knc.c:
```diff
@@ -1043,8 +1052,23 @@ move_data(work_t *work)
 					"read and local side write."));
-			shutdown_or_close(work->network_fd, SHUT_RD);
-			shutdown_or_close(work->local_out, SHUT_WR);
+			if (0 != shutdown(work->network_fd, SHUT_RD)) {
```
Ditto.

gstrauss replied:
ok.  modified.

gstrauss replied:
done.

gstrauss committed:
check for both EAGAIN and EWOULDBLOCK

nicowilliams commented on bin/knc.c:
```diff
@@ -1072,42 +1111,50 @@ move_data(work_t *work)
+		pfd_err->events = POLLRDHUP;
 
 		/*
 		 * We have this timeout only to allow us to recover
```
Ah, so maybe we could have a signal handler for `SIGCHLD` to write a single byte to a self-pipe we could add the read end of to the event loop, then we wouldn't need the timeout at all.

gstrauss replied:
The code already has a handler for SIGCHLD and ignores it.  The handler also does not set SA_RESTART.

Given that the program has at most one child that it creates, we could just set a volatile sig_atomic_t static var upon receiving SIGCHLD and check for that, not needing a self-pipe at all.

gstrauss replied:
If child dies, I think we should waitpid() in the signal handler for SIGCHLD and set a flag.  In move_data(), if that flag is set, then we should read and discard data from the network, and as soon as local buffers (read from child and sent to network) are flushed, then exit.

nicowilliams commented on bin/knc.c:
```diff
@@ -1072,42 +1111,50 @@ move_data(work_t *work)
 		 */
-		tv.tv_sec = 30;
-		tv.tv_usec = 0;
+		timeout = POLLRDHUP ? -1 : 30000; /* milliseconds */
```
See above comment about defaulting `POLLRDHUP` to `POLLHUP`.

nicowilliams replied:
Yeah, for now just leave the timeout as `30000`, but then, see above comment about catching `SIGCHLD`: if we do the signal handler self-pipe thing then we can have the `poll(2)` timeout be `0` always.

gstrauss replied:
I don't think we need a timeout if we handle SIGCHLD and take action to avoid blocking.  Similar to HTTP, we probably want to read and sink data from the network (up to some limit) to give the client a chance to read any pending response.

gstrauss replied:
If we are waiting for events on the sockets, whether or not we spawned a child on a socketpair() should not matter.  I think that the changes made along with the conversion to use poll() obviate the need for the timeout, as long as I also make a change to waitpid() the child in the signal handler.  We might instead choose to SIG_IGN for SIGCHLD, which is allowed in POSIX.1-2008.

gstrauss replied:
I pushed an additional commit which waitpid() for SIGCHLD in the signal handler, so that debug logging of the reaped child occurs.

nicowilliams commented on bin/knc.c:
```diff
@@ -999,40 +1007,41 @@ write_network_buffer(work_t *work)
+	pfd_net->fd = work->network_fd;
+	pfd_in->fd  = work->local_in;
+	pfd_out->fd = work->local_out;
+	pfd_err->fd = work->local_err;
```
~I think this needs to be done in the loop.  Yeah, one might think that we'd never want to continue if either side is fully closed, but in fact we do: to avoid losing buffered data.~

gstrauss committed:
retry (later) on read() or write() if EINTR,EAGAIN

fixes bug whereby interrupted read()/write() could corrupt data stream

nicowilliams commented:
BTW, knc supports half-close.  It needs to continue to do so (not that this breaks that -- I've not determined that yet).

gstrauss committed:
check shutdown flags run no more than once/sock

check that shut_nread_lwrite, shut_nwrite_lread run no more than once

gstrauss committed:
use SA_RESTART for smoother syscalls

use poll() and non-blocking accept() in do_listener() accept loop

gstrauss committed:
fix bug which misdetected error on read local sock

-Werror=type-limits

Due to signedness difference, move_local_to_network_buffer() misdetected
error on read() from local sock, so would end up aborting if any read()
error occurred (including EINTR or EAGAIN) and did not gracefully handle
ECONNRESET.  Using ssize_t instead of size_t allows for more graceful
handling.

gstrauss committed:
fix missing prototypes: -Werror=missing-prototypes

XXX: many of these routines could be declared 'static', too

gstrauss committed:
fix comparisons of different signedness

-Werror=sign-compare

gstrauss committed:
fix warning for pointer arithmetic on (void *)

-Werror=pointer-arith

gstrauss committed:
fix warning for discarding const

-Werror=cast-qual

gstrauss committed:
remove unused variables

-Werror=unused-variable

gstrauss committed:
move declaration for static func never defined

‘debug_printf’ declared ‘static’ but never defined

-Werror=unused-function

gstrauss committed:
enable -Wall -Werror for bin/* w/ enable-developer

./configure --enable-developer

gstrauss commented:
I broke the merge request into many commits and added commits to fix compiler warnings (and errors).

I think the fixes up until the final two commits should be a separate merge request, and the final two commits to convert knc to use poll().

As things stand, if the child exits, but the remote client is still sending data, we do not sink that data (read and discard), but we do try to send remaining data from child to the network (for the remote client) before exiting.  If the data sink is desirable, I think we should handle that separately, and we should probably implement a timeout before abandoning, as the socket send buffers will also be full.

gstrauss committed:
check for socket close if POLLRDHUP not supported

check for both network socket close and for local socket close
after timeout if most recent poll() did not ask for POLLIN events
on each socket

gstrauss committed:
reorder write_buffer_t to improve memory locality

gstrauss committed:
separate func for write_local_err()

loop to log up to 64k at a time from local err

gstrauss committed:
fix potential timeout or livelock after EPIPE

discard pending write buffers upon EPIPE

If pending data exists in 'in' buffers, then knc might never propagate
shutdown() after receiving EPIPE.

gstrauss committed:
remove option duplication in knc.1

gstrauss commented:
I split this into three different PRs:
https://github.com/gstrauss/knc/pull/2  knc bugs
https://github.com/gstrauss/knc/pull/3  knc cc warnings
https://github.com/gstrauss/knc/pull/4  knc use poll()

gstrauss committed:
allocate work_t after accept()

gstrauss committed:
check time limit at start of do_listener() loop

gstrauss committed:
fix potentially unaligned pointer dereference

replace dereference of type-punned pointer;
dereferencing type-punned pointer will break strict-aliasing rules

-Werror=strict-aliasing

gstrauss committed:
exit soon after -T <max_time> expires

nicowilliams commented on bin/knc.c:
```diff
@@ -955,7 +959,9 @@ write_network_buffer(work_t *work)
-	if (len < 0 && ((errno != EINTR) && (errno != EAGAIN))) {
+	if (len < 0) {
+		if (errno == EINTR || errno == EAGAIN)
+			return 1; /* retry later (do not return <= 0) */
```
The comment block for this function needs to be updated.  Callers don't treat `1` as "fully transmitted" but as "no error".

nicowilliams commented on bin/knc.c:
```diff
@@ -795,6 +795,8 @@ move_local_to_network_buffer(work_t *work)
 		return 0;
 	} else if (work->network_buffer.in_len < 0) {
+		if (errno == EINTR || errno == EAGAIN)
+			return 1; /* retry later (do not return <= 0) */
```
Add a comment block for this function documenting the return values.

nicowilliams commented on bin/gssstdio.c:
```diff
@@ -472,7 +472,7 @@ read_packet(int fd, gss_buffer_t buf, int timeout, int first)
 
 		if (ret == -1) {
-			if (errno == EINTR || errno == EAGAIN)
+			if (errno==EINTR||errno==EAGAIN||errno==EWOULDBLOCK)
```
Whitespace issues.

FYI, @elric1 is very particular about max 79 columns.

nicowilliams commented on bin/gssstdio.c:
```diff
@@ -472,7 +472,7 @@ read_packet(int fd, gss_buffer_t buf, int timeout, int first)
 
 		if (ret == -1) {
-			if (errno == EINTR || errno == EAGAIN)
+			if (errno==EINTR||errno==EAGAIN||errno==EWOULDBLOCK)
```
We might need a

```
#ifndef EWOULDBLOCK
#define EWOULDBLOCK EAGAIN
#endif
```

gstrauss replied:
As discussed, will add this later if we determine this is needed.

nicowilliams commented on bin/knc.c:
```diff
@@ -869,6 +869,7 @@ write_local_buffer(work_t *work)
 			 */
 			LOG(LOG_DEBUG, ("write got EPIPE"));
 
+			work->local_buffer.in_valid = 0;
```
Per your explanation, this fixes something of a livelock/timeout because we don't `shutdown()` if there's pending data.  Not sure this is the best way to fix that, but I'll take it, but I do want the commit comment to explain the observable behavior -- that this commit fixes a possible timeout.

nicowilliams commented on bin/knc.h:
```diff
@@ -59,11 +59,11 @@ extern prefs_t prefs;
 	char			in[GSTD_MAXPACKETCONTENTS];
 	char			in_valid;
-	size_t			in_len;
+	ssize_t			in_len;
```
This requires thinking.  I think these really should be `size_t` -- unsigned -- with negative `ssize_t` return values handled where they are produced.

nicowilliams replied:
Per-discussion, you went for this change to keep it tiny, but I'd rather we keep these as unsigned.

nicowilliams commented on bin/knc.c:
```diff
@@ -56,18 +56,25 @@ prefs_t prefs;
 /* BEGIN_DECLS */
 
 void	sig_handler(int);
+void	sig_set(int, void (*)(int), int);
```
Well, OK, but it's a single source code file program, so they should all be static.  But whatever.

gstrauss replied:
As discussed, I am not going to make this change at this time, but we might choose to do so in a later PR.

nicowilliams commented on bin/gssstdio.c:
```diff
@@ -244,8 +244,7 @@ gstd_accept(int fd, char **display_creds, char **export_name, char **mech)
 void *
-gstd_initiate(const char *hostname, const char *service, const char *princ,
-	      int fd)
+gstd_initiate(const char *hostname, const char *service, char *princ, int fd)
```
There's a better way.  We really are treating this as const, so we should declare it as const and below, we can unconst:

```
-		name.value = (char *) princ;
+		name.value = (void *)(uintptr_t)princ;
```

In Heimdal, for example, we have a `rk_UNCONST()` macro that does exactly this.

nicowilliams replied:
So do keep the const in the prototype.

nicowilliams commented on bin/gssstdio.c:
```diff
@@ -496,7 +496,7 @@ read_packet(int fd, gss_buffer_t buf, int timeout, int first)
 
 	/* We have the complete length */
-	len = ntohl(*(uint32_t *)len_buf);
+	len = ntohl(len);
```
The commit comment should be something like `Fix potentially unaligned pointer dereference`.

nicowilliams replied:
Commit subject lines need to be pithy, yes, but as much as possible they need to describe the bug being fixed (if a bug) so that `git log --oneline` can be helpful.

gstrauss reviewed (commented):
Comments have been addressed and branch force-pushed with the updates.

gstrauss commented:
I re-ordered some commits to move conversion from select() to poll() to be later.

gstrauss committed:
handle SIGCHLD separately after launch_program()

SIGCHLD handling of backend program is now separate from SIGCHLD
handling of accept() loop (move child reaping out of move_data()
to possibly allow for future replacement with libknc)

Fix bug whereby failure to fork() would incorrectly increment
num_children in do_listener()

gstrauss committed:
knc: set TCP_NODELAY on sockets (disable Nagle)

gstrauss committed:
knc: move duplicated code into so_keepalive_set()

elric1 commented:
@nico may have mentioned that I don't think that changing bin/knc.c from `select(2)` to `poll(2)` is terribly useful. I inherited the `select(2)` decision, I didn't write it. There are two reasons that I never decided to change it:

1. I plan to rewrite the code to use `libknc(3)` which will use `poll(2)`, any unnecessary churn of `bin/knc.c` is therefore wasted effort, and
2. `bin/knc.c` always has essentially the same fds at the same numbers and the numbers are small. `poll(2)` doesn't offer a huge advantage in this particular use case.

gstrauss commented:
The code is already written and tested.  The proper use of `poll(2)` `POLLRDHUP`, as well as `POLLHUP` and `POLLERR`, is an advantage over `select(2)`.

Even without the final three commits in this PR, the prior 23 commits, each quite small, fix bonafide bugs and compiler warnings and errors, so I hope they will be considered.  Having already been written and tested, this PR is far and away ahead of a future planned migration to use libknc.

gstrauss committed:
use _exit() in child instead of exit()

gstrauss committed:
adjust emit_key_value() check for '\n' and comment
